### PR TITLE
Broker #5: Move initialization out of the constructor

### PR DESCRIPTION
**Pull request by tnorling**
Moves broker initialization code out of the constructor and into a new public API. 
May make the initialize function generic prior to release to provide a method for other non-broker initialization code as well.

### CODE REVIEW

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 1]
```diff
@@ -46,21 +47,19 @@ export class PublicClientApplication extends ClientApplication implements IPubli
...
         super(configuration);
-        this.initializeBrokering();
     }
 
     /**
      * 
      */
-    private initializeBrokering(): void {
-        if (this.config.system.brokerOptions.actAsBroker) {
+    async initializeBrokering(): Promise<void> {
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/PublicClientApplication.ts:55]: We should add this to `IPublicClientApplication` as well.

File: samples/msal-browser-samples/BrokerTestApp2.0/app/brokeredApp/auth.js [hunk 2]
```diff
@@ -11,9 +13,12 @@ const isEdge = msedge > 0;
...
 let username = "";
 
-// Create the main myMSALObj instance
-// configuration parameters are located at authConfig.js
+/*
+ * Create the main myMSALObj instance
+ * configuration parameters are located at authConfig.js
+ */
 const myMSALObj = new msal.PublicClientApplication(msalConfig);
+myMSALObj.initializeBrokering();
```
Comment (jasonnutter) [inline samples/msal-browser-samples/BrokerTestApp2.0/app/brokeredApp/auth.js:21]: We should show this being awaited or .thened.

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 3]
```diff
@@ -46,21 +47,19 @@ export class PublicClientApplication extends ClientApplication implements IPubli
...
+        if (this.config.system.brokerOptions.actAsBroker && !BrowserUtils.isInIframe()) {
             this.broker = new BrokerClientApplication(this.config);
+            this.broker.listenForBrokerMessage();
         } else if (this.config.system.brokerOptions.allowBrokering) {
             this.embeddedApp = new EmbeddedClientApplication(this.config, this.logger, this.browserStorage);
             this.logger.verbose("Acting as child");
-            this.embeddedApp.initiateHandshake().catch((e) => {
-                this.logger.error(`Broker handshake failed: ${e}`);
-            });
+            await this.embeddedApp.initiateHandshake();
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/PublicClientApplication.ts:62]: Log messages would be good here, including a warning if both flags are enabled and in they are in the top frame.

File: samples/msal-browser-samples/BrokerTestApp2.0/app/brokeredApp/auth.js (hunk 2 above)
Comment (tnorling) [reply samples/msal-browser-samples/BrokerTestApp2.0/app/brokeredApp/auth.js:21]: Agreed. Can't await on the top level so we'll need to use .then and probably enable the login buttons inside

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 1 above)
Comment (tnorling) [reply lib/msal-browser/src/app/PublicClientApplication.ts:55]: Done

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 3 above)
Comment (tnorling) [reply lib/msal-browser/src/app/PublicClientApplication.ts:67]: Done

File: samples/msal-browser-samples/BrokerTestApp2.0/app/brokeredApp/auth.js (hunk 2 above)
Comment (tnorling) [reply samples/msal-browser-samples/BrokerTestApp2.0/app/brokeredApp/auth.js]: Done

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 4]
```diff
@@ -46,21 +47,24 @@ export class PublicClientApplication extends ClientApplication implements IPubli
...
 
     /**
      * 
      */
-    private initializeBrokering(): void {
-        if (this.config.system.brokerOptions.actAsBroker) {
+    async initializeBrokering(): Promise<void> {     
+        if (this.config.system.brokerOptions.actAsBroker && !BrowserUtils.isInIframe()) {
+            if(this.config.system.brokerOptions.allowBrokering) {
+                this.logger.verbose("Broker options actAsBroker and allowBrokering both set to true. Preferring to act as broker.");
```
Comment (jmckennon) [inline lib/msal-browser/src/app/PublicClientApplication.ts:58]: What does "preferring to act as broker" mean?

**Review by jmckennon — APPROVED** [review]
looks good, just had a question about logger message/wording.

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 4 above)
Comment (tnorling) [reply lib/msal-browser/src/app/PublicClientApplication.ts:58]: It just means that there are conflicting config options so we are silently respecting the `actAsBroker` flag and ignoring the `allowBrokering` flag. I suppose wording it simply `Acting as broker.` would get the point across as well.

Comment (sameerag) [reply lib/msal-browser/src/app/PublicClientApplication.ts:58]: Does this imply it is either or and never both? If so can we spell it out?

Comment (tnorling) [reply lib/msal-browser/src/app/PublicClientApplication.ts]: Updated for clarity.

**Review by pkanher617 — APPROVED** [review]

**Review by jasonnutter — APPROVED** [review]

**Pull request merged**